Zend/zend_API.c

```diff
@@ -1682,7 +1682,15 @@ ZEND_API int array_set_zval_key(HashTable *ht, zval *key, zval *value) /* {{{ */
 			result = zend_hash_index_update(ht, Z_LVAL_P(key), value);
 			break;
 		case IS_DOUBLE:
-			result = zend_hash_index_update(ht, zend_dval_to_lval(Z_DVAL_P(key)), value);
+			if (ZEND_DOUBLE_FITS_LONG(Z_DVAL_P(key))) {
+				result = zend_hash_index_update(ht, zend_dval_to_lval(Z_DVAL_P(key)), value);
+			} else {
+				zval tmp;
+				ZVAL_COPY(&tmp, key);
+				convert_to_string(&tmp);
```

Comment:
Wouldn't this mean that for floats < PHP_INT_MIN or > PHP_INT_MAX, they would be converted to strings with the fractional portion intact? That feels inconsistent with how we handle other floats. Maybe they should be truncated (rounded towards zero) before being converted…? Hmm :/

Reply:
Also, I'm not sure if `"NAN"` and `"INF"` keys make sense or not.

Reply:
@hikari-no-yume thanks! sounds reasonable to truncate it. But to be fair I concentrated on alternative solution here https://github.com/php/php-src/pull/2677

Reply:
Come to think of it actually, this depends on whether you're on 32-bit or 64-bit. For 64-bit systems, anything in that range can't have a “fraction” (the intervals between values are at least 2) so this point is moot there. For 32-bit there could be a fraction.

But anyway, yes, the alternative solution is probably better, I agree. :)